### PR TITLE
Feature new schema

### LINKED ISSUE #26: Team initialisation (Battle)
- Randomise players into red and blue teams
- username shd be their respective colors

### LINKED ISSUE #29: Schema (Monster and BattleRoomState)
### Possible Schemas
**Monster:**
- players tackling it
- health
- behaviour/anims

**Team:**

**Question:**
- difficulty
- answer

**Player:**
- score
- health
- username

**BattleRoomState:**
- time
- round 
- team and players (with team score)
- curRound score
- totalScore
- totalIndiv score (could be under player as well)
- allRounds count (num of rounds won)
- monsters (possibly location map)

### PR DESCRIPTION
**Pull request by winson8222**
- Added schema according to the new structure
- Integrated with BattleRoom and used schema to handle timer
1. For ArraySchema/MapSchema on server, use onAdd, onRemove as listener on Client side (room.state.players.onAdd(...))
2. For primitive type, use onChange as listener, (room.state.timeRemaining.onChange(...))
[](https://docs.colyseus.io/state/schema-callbacks/#onadd-fn-item-key-triggerall-true)
- tsconfig is changed to change tscompiler behavior to handle schema classes

### CODE REVIEW

File: .idea/shelf/Uncommitted_changes_before_Checkout_at_14_2_24__5_45_pm__Changes_.xml
Comment (shashahchk) [inline .idea/shelf/Uncommitted_changes_before_Checkout_at_14_2_24__5_45_pm__Changes_.xml:1]: can we remove these? Or should commit again

File: server/src/rooms/schema/Character.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  type,
+  MapSchema,
+  ArraySchema,
+} from "@colyseus/schema";
+
+export abstract class Character extends Schema {
+  @type("number") x: number;
+  @type("number") y: number;
+  @type("string") pos: string;
+  @type("string") lastMovedTime: string;
```
Comment (shashahchk) [inline server/src/rooms/schema/Character.ts]: note how this lastMovedTime is string, but the curRoundStartTime in BattleRoomState is number. Should try to be consistent?

File: server/src/rooms/schema/Character.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  @type("boolean") isMoving: boolean;
+  @type("number") id: number;
+}
+
+export class Player extends Character {
+  @type("string") userName: string;
+}
+
+export class Monster extends Character {
+  @type(["string"]) playersTackling = new ArraySchema<string>();
```
Comment (shashahchk) [inline server/src/rooms/schema/Character.ts]: should we name as playerIdsTackling?

File: server/src/rooms/schema/Character.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  @type("string") monsterType: string;
+}
+
+export class InBattlePlayer extends Player {
+  @type("number") health: number;
+  @type("number") totalScore: number;
+  @type("number") roundScore: number;
+  @type("number") teamId: number;
+  @type("boolean") isAlive: boolean;
+  @type(["number"]) questionsSolved: number; // abit sus but we'll see
```
Comment (shashahchk) [inline server/src/rooms/schema/Character.ts]: for all these names, can consider the use of id to differnetiate the Object from the ID,
in this case is it better to use questionIdsSolved?
consider this for all other fields as well

File: server/src/rooms/schema/Group.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+  type,
+  MapSchema,
+  ArraySchema,
+} from "@colyseus/schema";
+import { Monster, InBattlePlayer } from "./Character";
+
+export class BattleTeam extends Schema {
+  @type({ map: InBattlePlayer }) teamPlayers = new MapSchema<InBattlePlayer>();
+  @type("number") teamId: number;
+  @type("string") teamColor: string;
```
Comment (shashahchk) [inline server/src/rooms/schema/Group.ts]: consider using enum for more safety.

File: server/src/rooms/BattleRoom.ts [hunk 5]
```diff
@@ -81,27 +95,29 @@ export class BattleRoom extends Room<MyRoomState> {
...
     // (client.sessionId is unique per connection!)
+    this.state.teams[player.teamId].teamPlayers.set(client.sessionId, player);
     this.state.players.set(client.sessionId, player);
   }
 
   onLeave(client: Client, consented: boolean) {
     console.log(client.sessionId, "left!");
 
-    this.state.players.delete(client.sessionId);
+    this.state.teams[0].teamPlayers.delete(client.sessionId);
```
Comment (shashahchk) [inline server/src/rooms/BattleRoom.ts]: try not to hardcode this. 
There should be a way to retrieve the InBattlePlayer from the sessionId (can get this from the map in room state)
From that, we retrieve player.teamId,
then delete accordingly

File: server/src/rooms/schema/BattleRoomState.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+export class BattleRoomState extends MyRoomState {
+  @type([BattleTeam]) teams = new ArraySchema<BattleTeam>();
+  @type({ map: Monster }) monsters = new MapSchema<Monster>();
+  @type("number") totalRounds: number;
+  @type("number") currentRound: number;
+  @type("number") roundDurationInMinute: number;
+  @type("number") currentRoundTimeRemaining: number;
+  @type("number") roundStartTime: number;
+  @type("string") currentGameState: string;
```
Comment (shashahchk) [inline server/src/rooms/schema/BattleRoomState.ts]: can consider using enum for more safety

File: server/src/rooms/schema/MyRoomState.ts [hunk 7]
```diff
@@ -1,13 +1,13 @@
...
+  type,
+  MapSchema,
+  ArraySchema,
+} from "@colyseus/schema";
+import { Monster, Player } from "./Character";
+import { BattleRoom } from "../BattleRoom";
+import { BattleTeam } from "./Group";
 
 export class MyRoomState extends Schema {
   @type({ map: Player }) players = new MapSchema<Player>();
```
Comment (shashahchk) [inline server/src/rooms/schema/MyRoomState.ts:11]: this is the map i am referring to that maps sessionId/playerId to player Object. I'm not entirely sure on this. But this should allow u to handle the deletion of player when they leave battle nicely.

**Review by pygmychameleon — CHANGES_REQUESTED** [review]
Thanks a lot for the detailed schema and for updating the rest of the code to adhere to the schema. 
Just a few concerns about making certain types stricter to avoid problems in the future. These should be easy fixes. Also one main thing (deletion of player) that should be implemented correctly.

**Comment by yongning0310** [comment]
please pull and run the code before merging, im not sure if its correct cos i had to resolve merge conflict for parts of code that idek whats gg on

**Comment by yongning0310** [comment]
fix animation bug too

**Review by pygmychameleon — CHANGES_REQUESTED** [review]
(to yongning) Direction and animations works well. There is collision box bug. Namely when press left, probably due to the offset, (find out more about phaser physics collider). Also collision box for faune is sometimes not consistent size.

**Pull request merged**